### ISSUE TITLE
"ObjC Class 'b'WrappedPyObject'' couldn't be found." when running on iOS device

### ISSUE THREAD

**Issue by Skifarer** [beeware/rubicon-objc#226]
Trying to run a clean app on real iOS device, the `ObjCClass("WrappedPyObject")` line fails in rubicon's `api.py`, also causing a failure in the `add_method` function before the app loads.

Created a brand new app from `briefcase new`, `briefcase create iOS`, and `briefcase build iOS`, and then opened the .xcodeproj file in Xcode. Building the app and running it on a simulated iPhone 13 Pro with iOS 15.5 worked fine. However, when trying to run the same app on a real iPhone of this spec, the errors below occurred.

I had to sign the app with my own personal team, and I'm not sure whether this is even an acceptable thing to do with the Xcode project file - please delete if not!!

MacBook Pro (13-inch, M1, 2020) macOS Monterey 12.3.1
Xcode 13.4
iPhone 13 Pro iOS 15.5
Briefcase 0.3.7
rubicon_objc-0.4.2
Python version: 3.10.4 
Build: b2
libFFI: 3.4.2
BZip2: 1.0.8
OpenSSL: 1.1.1o
XZ: 5.2.5

```
Traceback (most recent call last):
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 1683, in <module>
    WrappedPyObject = ObjCClass("WrappedPyObject")
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 1113, in __new__
    ptr, name = cls._new_from_name(name_or_ptr)
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 985, in _new_from_name
    raise NameError("ObjC Class '%s' couldn't be found." % name)
NameError: ObjC Class 'b'WrappedPyObject'' couldn't be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "Hello World.app/Library/Application Support/com.example.helloworld/app/helloworld/__main__.py", line 4, in <module>
    main().main_loop()
  File "Hello World.app/Library/Application Support/com.example.helloworld/app/helloworld/app.py", line 27, in main
    return HelloWorld()
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/toga/app.py", line 272, in __init__
    self.factory = get_platform_factory(factory)
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/toga/platform.py", line 38, in get_platform_factory
    from toga_iOS import factory
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/toga_iOS/factory.py", line 1, in <module>
    from .app import App, MainWindow
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/toga_iOS/app.py", line 3, in <module>
    from rubicon.objc import SEL, objc_method
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/__init__.py", line 16, in <module>
    from . import api  # noqa: F401
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 1685, in <module>
    class WrappedPyObject(NSObject):
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 1119, in __new__
    ptr, name, attrs = cls._new_from_class_statement(name_or_ptr, bases, attrs, protocols=protocols)
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 1049, in _new_from_class_statement
    class_register(ptr, attr_name)
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/api.py", line 541, in class_register
    add_method(class_ptr, name, self, self.encoding)
  File "Hello World.app/Library/Application Support/com.example.helloworld/app_packages/rubicon/objc/runtime.py", line 911, in add_method
    imp = cfunctype(method)
MemoryError
```

**Comment by Skifarer** [comment]
Apologies, also:
Python version: 3.10.4 
Build: b2
---------------------
libFFI: 3.4.2
BZip2: 1.0.8
OpenSSL: 1.1.1o
XZ: 5.2.5

**Comment by freakboy3742** [comment]
Ugh... thanks for the report. Looks like this slipped through my release testing; I can reproduce the problem. I have no idea why it's happening, though. 

It looks like it isn't specific to Python 3.10; I can reproduce on 3.9 and 3.8. My immediate thought was that it might be related to some of the template changes related to PyConfig initialisation, or the weak memory reference handling stuff introduced in Rubicon-Objc 0.4.2; but reverting to an old template and Rubicon version doesn't appear to resolve the issue either. 

Looks like I've got some fun debugging ahead of me :-(

**Comment by freakboy3742** [comment]
Would be interested in your thoughts @dgelessus @samschott if you have any spare time...

**Comment by freakboy3742** [comment]
That's definitely the same end result (and the test case provided also works in this case), but I don't think it's the same root cause. 

That bug is reporting an issue with #defines resulting in the FFI closure not being allocated. We currently build the ctypes module with `-DHAVE_FFI_CLOSURE_ALLOC` but *not* `-DDUSING_APPLE_OS_LIBFFI, so we won't be affected in the same way as that bug report suggests.

The issue appears to be deeper in the `ffi_closure_alloc()` method - using the same example presented in that bug report, we *are* invoking `ffi_closure_alloc()`, but iOS devices return `None`, which then raises the MemoryError.

Digging way into the depths of libFFI, the underlying problem appears to be caused [by the call to `vm_remap` when creating the trampoline](https://github.com/libffi/libffi/blob/master/src/closures.c#L233). On the simulator, `cur_prot = VM_PROT_READ | VM_PROT_EXECUTE`. On device, `cur_prot = VM_PROT_READ`, so the VM_PROT_EXECUTE check fails, and so does the whole trampoline creation.

The (barely) good news is that if you comment out the VM_PROT_EXECUTE check, ctypes *appears* to work well enough for the purposes of Rubicon and Toga... but I have *absolutely* no idea if this is a good idea or not, and Apples documentation of `vm_remap`... [exists 😡](https://developer.apple.com/documentation/kernel/1585336-vm_remap).

The closest I found to anything referencing this is [this thread on Twitter](https://twitter.com/Jakeashacks/status/1254488744771686401?s=20&t=yYmtdsQLqR_VBi20cdTTSA) which suggests that it might be related to iOS 13, and suggests a possible mitigation that makes some sort of sense under the circumstances.

**Comment by freakboy3742** [comment]
I've got a possible fix submitted as a pull request for libFFI: https://github.com/libffi/libffi/pull/718

**Comment by samschott** [comment]
Uh, that's ended up being quite a rabbit hole of a problem!

**Comment by freakboy3742** [comment]
Still waiting on libFFI (and hopefully someone from Apple) to review the patch; but I had a discussion with a friend who knows the internals of BSD reasonably well, and he wasn't surprised by the need for this patch, or the fact that it was device specific. When running on M1 simulator, you're running on the Darwin OS kernel, so memory protection will be more lax. On device, you're using the iOS kernel, and Apple has been ratcheting up the security provisions for some time. Marking memory as non-executable is part of that. It's also consistent with the fact that the problematic check was introduced as part of the patch for arm64e support in libFFI ([a patch that was contributed by an Apple employee](https://github.com/libffi/libffi/pull/565))

Closing as the fix was actually needed in libffi, and has been [included as part of the the support libraries](https://github.com/beeware/Python-Apple-support/commit/e09a9c6699fa6868cd651cd43580a9d1a74375ad); that fix has been released inin 3.10-b3, 3.9-b7, and 3.8-b9.
